azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunner.java

```diff
@@ -905,9 +916,32 @@ private void prepareJobProperties(final ExecutableNode node) throws IOException
       }
     }
 
+    // 6. If there are any node-specific flow overrides, we also apply them
+    final Map<String, Map<String, String>> nodeParams = this.flow.getExecutionOptions()
+        .getNodeParameters();
+    if (nodeParams != null && !nodeParams.isEmpty()) {
+      applyNodeOverrides(node, nodeParams, props);
+    }
+
     node.setInputProps(props);
   }
 
+  private void applyNodeOverrides(final ExecutableNode node,
+      final Map<String, Map<String, String>> nodeParams, final Props props) {
+    if (node.getParentFlow() != null) {
+      // apply recursively top->down
+      applyNodeOverrides(node.getParentFlow(), nodeParams, props);
+    }
```

Comment:
> which makes it a non backward compatible change

Hmm I don't see that? Node overrides is an entirely new feature, and this recursion only applies possible node overrides.

Are you suggesting that node overrides should somehow be aware of whether a property already exists on the node at hand? If so, I'm not sure what the rules for that would be. As far as I can see it would mean that this feature would loose the power of overriding existing properties. And it wouldn't be so useful any more.

I would leave this like it is implemented here:
- Flow overrides don't override existing properties (unless configured otherwise).
- Node overrides override props on the node in question and its children. The most specific node override wins.

Reply:
To repro please modify the code in `testNodeOverrides()` to include below block when checking job4 props:

 ```
final Props job1GeneratedProps = new Props();
 job1GeneratedProps.put("props4", "gjob4");
 InteractiveTestJob.getTestJob("innerflow:job1").succeedJob(job1GeneratedProps);
 assertStatus(flow, "innerflow:job4", Status.RUNNING);

 final Props job4Props = nodeMap.get("innerflow:job4").getInputProps();
 Assert.assertEquals("someValue", job4Props.get("props4"));
```
According to existing rules `props4` should be `gjob4`, however it gets evaluated to `innerflow-val-4` which is coming from the _nodeOverride_ to nested flow `innerflow`.

Reply:
If I comment out the recursion it works as expected: output props win.

Reply:
Sorry but I couldn't figure out how to include the code that you pasted above. And I don't see "someValue" generated anywhere. Could you clarify that?

Regarding the issue itself, I have totally overlooked how output props are applied, because I have never used that feature myself. But still, why should output props win over overrides? If there is a job override that specifically targets a node, why would it not override also output props? If user wants output props to be used, then they should not set a node override that applies to that given job. But I could be missing some other point here.

Reply:
Please, try running:
```
@Test
  public void testNodeOverrides() throws Exception {
    this.testUtil = new FlowRunnerTestUtil(EXEC_FLOW_DIR, this.temporaryFolder);

    final HashMap<String, String> flowProps = new HashMap<>();
    flowProps.put("props7", "execflow7");
    flowProps.put("props6", "execflow6");
    flowProps.put("props5", "execflow5");

    // enable overriding also for existing job props
    final FlowRunner runner = this.testUtil.createFromFlowMap(FLOW_NAME, null, flowProps,
        Props.of(ConfigurationKeys.EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED, "true"));
    // Set some node-specific overrides
    runner.getExecutableFlow().getExecutionOptions().addAllNodeParameters(ImmutableMap.of(
        "job2", ImmutableMap.of("job-prop-2", "job2-val-2", "props6", "job2-val-6"),
        "innerflow", ImmutableMap.of("props6", "innerflow-val-6", "props4", "innerflow-val-4"),
        // overrides by nested job id: this is the most specific, so always wins
        "innerflow:job4", ImmutableMap.of(
            "props14", "innerflow-job4-val-4", "props5", "innerflow-job4-val-5"),
        // overrides by plain job id: most specific after full nested id
        "job4", ImmutableMap.of(
            "props14", "job4-val-4", "props5", "job4-val-5", "props7", "job4-val-7")
    ));
    final Map<String, ExecutableNode> nodeMap = new HashMap<>();
    createNodeMap(runner.getExecutableFlow(), nodeMap);
    final ExecutableFlow flow = runner.getExecutableFlow();

    // 1. Start flow. Job 2 should start
    FlowRunnerTestUtil.startThread(runner);
    assertStatus(flow, "job2", Status.RUNNING);

    final Props job2Props = nodeMap.get("job2").getInputProps();
    Assert.assertEquals("shared1", job2Props.get("props1"));
    Assert.assertEquals("job2", job2Props.get("props2"));
    Assert.assertEquals("moo3", job2Props.get("props3"));
    Assert.assertEquals("execflow7", job2Props.get("props7"));
    Assert.assertEquals("execflow5", job2Props.get("props5"));
    // should've been overridden by nodeOverride
    Assert.assertEquals("job2-val-6", job2Props.get("props6"));
    Assert.assertEquals("shared4", job2Props.get("props4"));
    Assert.assertEquals("shared8", job2Props.get("props8"));
    // entirely new prop via nodeOverride
    Assert.assertEquals("job2-val-2", job2Props.get("job-prop-2"));

    InteractiveTestJob.getTestJob("job2").succeedJob();
    assertStatus(flow, "innerflow:job1", Status.RUNNING);

    final Props job1Props = nodeMap.get("innerflow:job1").getInputProps();
    Assert.assertEquals("job1", job1Props.get("props1"));
    Assert.assertEquals("job2", job1Props.get("props2"));
    Assert.assertEquals("job8", job1Props.get("props8"));
    // nodeOverride by the sub-flow parent
    Assert.assertEquals("innerflow-val-6", job1Props.get("props6"));
    Assert.assertEquals("execflow5", job1Props.get("props5"));
    Assert.assertEquals("execflow7", job1Props.get("props7"));
    Assert.assertEquals("moo3", job1Props.get("props3"));

    InteractiveTestJob.getTestJob("innerflow:job1").succeedJob();
    final Props job1GeneratedProps = new Props();
    job1GeneratedProps.put("props4", "gjob4");
    InteractiveTestJob.getTestJob("innerflow:job1").succeedJob(job1GeneratedProps);
    assertStatus(flow, "innerflow:job4", Status.RUNNING);

    final Props job4Props = nodeMap.get("innerflow:job4").getInputProps();
    Assert.assertEquals("job8", job4Props.get("props8"));
    Assert.assertEquals("job9", job4Props.get("props9"));
    // nodeOverride by the sub-flow parent
    Assert.assertEquals("innerflow-val-6", job4Props.get("props6"));
    // nodeOverride with plain job id
    Assert.assertEquals("job4-val-7", job4Props.get("props7"));
    // nodeOverride with nested id
    Assert.assertEquals("someval", job4Props.get("props4"));
    Assert.assertEquals("innerflow-job4-val-5", job4Props.get("props5"));
    Assert.assertEquals("shared1", job4Props.get("props1"));
    Assert.assertEquals("shared2", job4Props.get("props2"));
    Assert.assertEquals("moo3", job4Props.get("props3"));
  }
```
In the output you should see:
```
azkaban.execapp.FlowRunnerPropertyResolutionTest > testNodeOverrides FAILED
    org.junit.ComparisonFailure: expected:<[someval]> but was:<[innerflow-val-4]>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at azkaban.execapp.FlowRunnerPropertyResolutionTest.testNodeOverrides(FlowRunnerPropertyResolutionTest.java:144) 

```

`someval` value is just to indicate  the Assert that failed, in this case `props4` of `job4`. According to existing rule `job4->job1.output->innerflow->job2.output->flow.overrides->job4 shared props` the value should have been `gjob4 ` but it was the override in node `innerflow`. 

Here node overrides are not only overriding the values of properties in the .flow/.job files but they are also interfering with the precedence by level (first the job, then output of its input nodes and so on). If a user wants to override the value of an output property at runtime then they just need to set that property on the node that goes first in the by level precedence, the adjacent input node, but it shouldn't be the wrapping subflow node. That is my understanding of what's already implemented.

Reply:
Please note that code and output above is against initial commits, it hasn't been tested against latest updates like the commit using `Props` hierarchy vs doing `putAll`.

Reply:
This is maybe analogous with the question of whether flow override props should override existing job props or not (https://github.com/azkaban/azkaban/pull/2632). It's a matter of choice how the order of precedence is defined.

Just to be sure that I understand your point of view, here's a screenshot of the flow that we're talking about:

<img width="257" alt="execpropstest-job3" src="https://user-images.githubusercontent.com/4446608/108420241-3653d500-723c-11eb-9e7d-2528cbf4a7ef.png">

And so, we're looking especially at the case of resolving props of `innerflow:job4`.

The competing props sources are:
- properties file `subdir/shared.properties`, value: `shared4 `
- node-specific override of the parent flow `innerflow`, value: `innerflow-val-4`
- job output props of the previous job `innerflow:job1`, value: `gjob4`

It's obvious that the shared props never wins. But what comes after that is debatable IMO.

I understood that you would like to make job output props take precedence over node-specific overrides unless the node-specific override is specifically for the node in question.

Previously I thought that you had found a bug, where the node overrides are somehow affecting even some nodes above the node where it's set or something, but that doesn't seem to be the issue.

The scope that I intended for node-specific overrides is indeed that they override the node itself + every node that is inside the sub-flow, if the node is a flow. To me it's correct that node overrides are what their name says: overrides.

If user wants output props to be used, then they should NOT set a node override that applies to the job that is supposed to read those output props.

To me it doesn't seem intuitive that "level" of property carries higher weight in case of overrides. To me that sounds like then props of job4 itself should be more significant than a node override that was made on it's parent node `innerflow`.

Indeed you wrote:

> If a user wants to override the value of an output property at runtime then they just need to set that property on the node that goes first in the by level precedence, the adjacent input node, but it shouldn't be the wrapping subflow node.

But I don't think that would be intuitive.

When you're starting a flow and preparing to set node overrides, it's natural to think that ok, I want to target `innerflow` and everything under it, forcing the property to value `innerflow-val-4` regardless of what is set in existing props hierarchy, or any runtime output props. Now, it's also easy to make the choice of NOT targeting some node: just pick the nodes more specifically to only override props for those that you really want to. It would be less intuitive that one needs to carefully analyze which jobs are passing output props and which are not, in order to know how much you can leverage simple override on an entire flow node and how much you need to separately override individual jobs. In this case, with the order of precedence that you described, the `innerflow:job1` would not need that kind of exact override, but it would be needed on `innerflow:job4` because otherwise it would get the value from output props of `innerflow:job1`. If there are more nodes inside a subflow it can quickly get tricky to follow where output props would be used and where not. And it would typically mean having to set more lines of overrides, which is already tedious enough to do in the UI.

For the record, in my org we're not really using job output props anywhere, so I don't have a major issue with either way of defining the precedence for them. But if we would be using output props, I feel like I would have a harder time educating users that there's a special case where node overrides don't apply (ie. if override comes from parent & output props also have that key). It seems simpler to just say that override is the one that wins, so don't pick a scope that overrides something that you don't want overridden.

----

One more thing that I thought about, now that we're talking about this: this feature doesn't now allow setting overrides solely on a flow node  – the override always propagates to everything under it. But I think flow nodes are not really "executed" per se, so this is ok. Otherwise we would need to for example choose that overrides only affect the node itself (which would mean that override needs to be set on every node separately) OR have a toggle for "is transitive" on node overrides to make it optional OR even allow setting a number on each node override to specify how many levels deep to apply.

Sorry for the delay in answering this.

Reply:
> The scope that I intended for node-specific overrides is indeed that they override the node itself + every node that is inside the sub-flow, if the node is a flow. To me it's correct that node overrides are what their name says: overrides.

Ok, let's think about this scenario: A user has 5 jobs which take a _date_ property as parameter, 4 of them use the current date as value and the remaining job, because is a backfill, uses a date from last month. So, the user analyses that for the 4 jobs they will define the _date_ property at flow level (only), for convenience and to do it just once, and for the particular backfill job they will define it in the job itself which is the most specific place. Now, do you think that semantic(4 similar jobs, 1 different) will change when running the flow? I would say no. 
Let's say the user needs to override the value of the _date_ property at runtime for some reason, there are two behaviors for overrides that can apply:
1-Overrides apply only on the node.
2-Overrides apply on the node and everything under it.   

If 2) is the one implemented then the user will not only update current date for the for 4 similar jobs but their particular backfill job will get impacted as well which was not the original intention.


Regarding:
> When you're starting a flow and preparing to set node overrides, it's natural to think that ok, I want to target innerflow and everything under it, forcing the property to value innerflow-val-4 regardless of what is set in existing props hierarchy, or any runtime output props.
I don't think is as simple as override and forget, this will lead to unexpected behavior and outputs. To do a runtime override the user should know their workflow and the impact of their override, therefore they necessarily need to be aware of what output/inner-flow/job specific properties are in use and by what nodes. So I don't see this as a strong advantage of the "override everything below" behavior and I don't think users can avoid the analysis hassle behind: 
> If there are more nodes inside a subflow it can quickly get tricky to follow where output props would be used and where not.

As I see it both approaches (1-Overrides apply only on the node and 2-Overrides apply on the node and everything under it) have advantages and one can be more intuitive/useful than the other one under certain scenarios so it comes down to picking one I guess. However, given that the precedence rule `job4->job1.output->innerflow->job2.output->flow.overrides->job4 shared props` which exists today will break (or would not apply in all cases) if we merge your code as is, I would kindly ask you to put behind a feature flag (maybe reuse EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED) the "override everything below" behavior and keep the "override only at node" as the default to avoid possible issues for users making use of that rule. At LinkedIn we do have users leveraging the job output properties.

Apologies also for the delayed response! First I was on vacation for a week and after I came back things have been very busy.

Reply:
@ypadron-in, thanks for all the time that you're taking to review this. I am not able to have super quick response either. I'm comfortable as long as I can feel that this PR is moving forward (even if gradually), and is not at risk of being abandoned :)

> therefore they necessarily need to be aware of what output/inner-flow/job specific properties are in use and by what nodes

Agreed, this burden is always on the user. User just needs to know how the overrides behave to be able to successfully make use of them to achieve the desired outcome.

> I would kindly ask you to put behind a feature flag (maybe reuse EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED) the "override everything below" behavior and keep the "override only at node" as the default

I thought the behavior that makes sense in your setup would be to "override below unless defined". I thought you would also think it makes sense that overrides on a subflow node affect the jobs inside it – unless there's a conflict ie. either the inside job itself defines the same property, or there's an output property by that name from the previous job. Is it so that you would want to treat node specific overrides always node specific? That would basically mean that there is no meaning to ever target a subflow node with a node-specific override, only actual leaf nodes.

I try to clarify the rules of precedence & propagation that I'm proposing below. Here I'm assuming that you would want "override below unless defined" instead of "override only at node" – but please let me know which one you prefer. I think it "override only at node" is selected, then you would not want to see the subflow node ids in the dropdown either, only let job node ids.

Any way, let's consider this flow as an example:

- subflow: type=flow, flow.name=innerJobB
- innerJobA: type=job, defines prop.y=definedByJobA, writes output property prop.x=aOutputProp
- innerJobB: type=job, depends on innerJobA, defines property prop.y=definedByJob

Now, user sets these node overrides:
- subflow: prop.x=subFlowX
- subflow: prop.y=subFlowY

Then the resolved props would be..

If EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED=false (default / LinkedIn conf):

node | resolved props
-- | --
subflow | prop.x=subFlowX, prop.y=subFlowY
subflow:innerJobA | prop.x=subFlowX, prop.y=definedByJobA
subflow:innerJobB | prop.x=aOutputProp, prop.y=definedByJob

- The only "interesting" case where override was applied was `prop.x` for `subflow:innerJobA`

If EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED=true:

node | resolved props
-- | --
subflow | prop.x=subFlowX, prop.y=subFlowY
subflow:innerJobA | prop.x=subFlowX, prop.y=subFlowY 
subflow:innerJobB | prop.x=subFlowX, prop.y=subFlowY

How to have the same outcome as the first example if EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED=true? Then user could set this:

Node overrides:
- subflow:innerJobA: prop.x=subFlowX

node | resolved props
-- | --
subflow | 
subflow:innerJobA | prop.x=subFlowX, prop.y=definedByJobA
subflow:innerJobB | prop.x=aOutputProp, prop.y=definedByJob

So, it's not exactly the same. But in practice I at least don't know that the props for a _subflow_ node itself have any meaning. So not being "able" to set overrides only on a subflow node doesn't mean that user is missing some functionality.

And then the other way around: how to have the same result as the second example, if EXECUTOR_PROPS_RESOLVE_OVERRIDE_EXISTING_ENABLED=false (default / LinkedIn conf)? By setting everything specifically:

Node overrides:
- subflow:innerJobA: prop.x=subFlowX
- subflow:innerJobA: prop.x=subFlowX, prop.y=subFlowY 
- subflow:innerJobB: prop.x=subFlowX, prop.y=subFlowY 

node | resolved props
-- | --
subflow | prop.x=subFlowX, prop.y=subFlowY
subflow:innerJobA | prop.x=subFlowX, prop.y=subFlowY 
subflow:innerJobB | prop.x=subFlowX, prop.y=subFlowY

Although I'm not 100% sure how you would want the output property to be treated. Do you think the above makes sense? ie. output property gets ignored if a node-specific override exists for the job that would read that output property.

Could you confirm if these examples match what you had in mind for how this should work, or if needed adjust the example cases as desired?

Thank you for allowing optional behavior driven by a feature flag.

Reply:
Thank you for being so patient!
About:
> I thought the behavior that makes sense in your setup would be to "override below unless defined". I thought you would also think it makes sense that overrides on a subflow node affect the jobs inside it – unless there's a conflict ie. either the inside job itself defines the same property, or there's an output property by that name from the previous job.

Your understanding is correct, this paragraph accurately describes what I meant, sorry if my wording wasn't clear.
The examples you provide also comply with existing props resolution rules which is what I'm ultimately looking for: retaining the ability to apply existing behavior.